Add ability to bind arbitrary functions.

TAAPArthur opened the pull request:
Before all the predated commands where kept in an array and their
indexes were used in bindings. This meant that users couldn't add their
own functions from the config file. Now key/mouse bindings have been
changed to to store the function ptr (wrapped in a cmd_t struct to also
store the mode) directly instead.

General cleanup done in this commit:
Defined `MODE_ALL` instead of using the magic number -1.
Made `imgcursor` a config variable.

For example, suppose one had bindings like:
```
{ 0,                   XK_q,             g_quit,                     None },
{ ShitMask,            XK_q,             {quit_err},                 None }
{ ControlMask,         XK_q,             {quit_err, .mode=ALL_MODE}, None }
```
The existing binding `q` has been left unchanged and is defined the same
way. However, the new hypothetical binding `Shift-q` can be used to call
the custom function quit_err when in IMAGE_MODE (default). Similary
`Ctrl-q` can to the same in any mode.

eylles commented:
i can see the benefit from this, altho the changes needed seems somewhat drastic at a first glance.

N-R-K commented:
If the goal is to call an existing function or define your own custom functions, then it can be done via

* defining a wrapper/custom function in `commands.c`
* then listing it in `commands.lst`.

This patch just makes it a bit more convenient by bypassing the above 2 steps. I'm not quite sure if that's worth making such a structural change for that, although I don't exactly oppose it either. But in case we do want to merge this, we should do it (and every other "big" changes) before `v28`.

TAAPArthur commented:
> This patch just makes it a bit more convenient by bypassing the above 2 steps.

I don't find that summary inaccurate... One of my goals is to be able to extend nsxiv from the config file without having to touch the main code base at least for features that are small 100% personal and have no change of getting into master. Right now I'm using it for define special functions for dual-page mode. I'm also needed for my post-installation configuration (which is basically as good as runtime configuration). Don't want to assume people care about my specific usecase, but I think the fact of making the code base more extensible is a plus. 

There's also something to be said about having all small probably related customization in one file.

N-R-K commented:
I've been thinking about this for a bit, if the goal is to be able to define custom functions inside `config.h` is there any other way to do that, without making such structural change?

Mostly just trying to see what our available options are and consider their pros/cons before proceeding with this.

TAAPArthur commented:
> I've been thinking about this for a bit, if the goal is to be able to define custom functions inside `config.h` is there any other way to do that, without making such structural change?

Yes, but I what I'm thinking of isn't cleaner. We could add 2 fields `bool (*func)(arg_t)` and `int arg` to arg_t and define a special builtin function that just runs `func(arg)` Believe this can have the same end result and the config syntax can be the same for users who won't want the ability (compiler may give warnings). For people who want to use custom functions, the keybinding struct is now a bit convoluted and different from other programs. I don't think making the diff smaller is worth it. 

What is the objection to make "structural changes"? Is it the risk of breaking patches? Or is it just the size of the diff? To me the big difference is that commands.lst and been transformed to commands.h. Technically we could keep the commands.lst and just tweak the sets of macros to the new format. The only benefit of doing so is that it becomes easier to index the predefined commands by name/number which could be useful for some ipc like interface. Even in that case, think it may be cleaner to just generate the commands list instead of relying or the preprocessor and having to include commands.lst multiple times.

N-R-K commented:
It's not about the diff at all. I'm mainly trying too see what other options we have before deciding on any.

TAAPArthur commented:
@eylles @N-R-K 
Looking at this again, I really don't think the change is as drastic/structable as it may appear. The "biggest" change was just inlining commands.lst instead of defing/undefining macros and including it twice. In exchange for being a hack it did reduce some redundancy. Can acheive the end goal with or without commands.lst so wouldn't want that to be what's causing the hesitation.

I don't think there is an alternative for anything else (important) in the commit. Getting rid of the fixed array and the index into the array in keybindings is the core point. There are ways to go about having user functions but think most would share this dependency.

N-R-K commented on commands.c:
```diff
@@ -63,6 +63,8 @@ bool cg_quit(arg_t _)
 		}
 	}
 	exit(EXIT_SUCCESS);
+	// silence complier warning
+	return 0;
```
Which compiler are you using? I don't get any warning on either GCC or Clang.

N-R-K commented on commands.c:
```diff
@@ -236,7 +238,7 @@ bool cg_navigate_marked(arg_t n)
 {
 	int d, i;
 	int new = fileidx;
-	
+
```
I think we've taken care of this in the "consistency" PR. Best to revert it here and keep the patch to functional changes only.

N-R-K replied:
I think we've taken care of this in the "consistency" PR. Best to revert it here and keep the patch to functional changes only.

N-R-K commented on main.c:
```diff
@@ -900,14 +897,6 @@ int main(int argc, char **argv)
 	filecnt = fileidx;
 	fileidx = options->startnum < filecnt ? options->startnum : 0;
 
-	for (i = 0; i < ARRLEN(buttons); i++) {
-		if (buttons[i].cmd == i_cursor_navigate) {
-			imgcursor[0] = CURSOR_LEFT;
-			imgcursor[2] = CURSOR_RIGHT;
-			break;
-		}
-	}
-
```
Wouldn't removing this cause issue?

TAAPArthur replied:
I just moved it into to config file so users had full control. Imagine a user wraps/extends i_cursor_navigate and use the new function for their mouse bindings? Would you expect your pointer icons to change?

N-R-K replied:
> I just moved it into to config file so users had full control.

Can you separate this into it's own commit then, we'll rebase the PR instead of squashing. I'd like to be able to have a rough idea of what a commit is changing just by reading the short commit msg on the commit log.

TAAPArthur replied:
Done

N-R-K commented on commands.c:
```diff
@@ -63,6 +63,8 @@ bool cg_quit(arg_t _)
 		}
 	}
 	exit(EXIT_SUCCESS);
+	// silence complier warning
+	return 0;
```
Which compiler are you using? I don't get any warning on either GCC or Clang.

eylles replied:
Perhaps he is using tcc or something similar, iirc just like all suckless tools sxiv can be compiled with tcc.

TAAPArthur replied:
Yes I was using tcc. Don't like getting into the habit of ignoring compiler messages so just added the (useless) line. Feel free to remove if it bothers you.

N-R-K commented:
> Can acheive the end goal with or without commands.lst so wouldn't want that to be what's causing the hesitation.

Let's keep commands.lst

eylles commented:
Perhaps the permanence or replacement of commands.lst for something else would be more fit for a different PR, perhaps something to address instances of duplicated code in the codebase?

TAAPArthur commented:
> > Can acheive the end goal with or without commands.lst so wouldn't want that to be what's causing the hesitation.
> 
> Let's keep commands.lst

Relooking at the code, keeping commands.lst isn't as clean as I had implied. It isn't impossible, but think it would change the config file (like going from `t_reload_all` to `T_CMD(reload_all)` in keybindings) or would extra complexity to the keybinding struct that I would classify as a hack. The limitation is do to not being able to define a macro in a macro and the lack of `constexpr`. One "simple" option is to auto generate a file based off of commands.lst. Could generate the needed file with 3 sed commands to commands.lst to convert it to the needed format... But don't think it is worth it unless we have some reason to want to keep commands.list. Apologies for the confusion.

N-R-K commented:
> Relooking at the code, keeping commands.lst isn't as clean as I had implied.

Then I guess we'll have to scratch that idea. @bakkeby Do you happen to have any idea on how else this can be achieved?

If no one else can come up with any other alternative, then I guess we'll have to move forward with this.

bakkeby commented:
@N-R-K no I do not see any reasonable way of keeping commands.lst. I am not entirely sure why we would want to anyway? Moving the definitions to `commands.h` seems like a more sane solution to me.

Personally I don't particularly like this dancing around just to hide the `MODE_ALL`/`MODE_IMAGE`/`MODE_THUMB` flags from the user as opposed to just having them in the keys array in `config.h`, but I can kind of understand why it was implemented like that in the first place.

TAAPArthur commented:
> Personally I don't particularly like this dancing around just to hide the `MODE_ALL`/`MODE_IMAGE`/`MODE_THUMB` flags

Completely agree. If we changed the config.def.h to inline the macros, I think that would be much cleaner. It would break existing configs and we seem to want to avoid that when possible so just kept the macros. I don't object to removing the macros and breaking config.

(Technically if we inlined the macros, we could keep commands.lst but don't think there is benefit in doing so)

N-R-K commented:
@TAAPArthur changed a couple things. Mainly moved imgcursor back to main since that's a bit out of scope for this PR. Also I don't think we needed MODE_ALL to be explicitly -1, so just using the default enum value for it instead.

Is there any benefit of guarding `command.h` with the `#ifdef` ? Other than that, I think I'm okay with this PR.

TAAPArthur commented:
> @TAAPArthur changed a couple things. Mainly moved imgcursor back to main since that's a bit out of scope for this PR. Also I don't think we needed MODE_ALL to be explicitly -1, so just using the default enum value for it instead.

So for future reference, I'd appreciate if changes that were made explicitly weren't just reverted. I welcome and appreciate the style nits, but there are subtle reasons for both of those changes. For imgcursor, if you wrap ci_cursor_navigate for whatever reason and use that in your bindings instead, your cursor changes. I don't think that is good or intuitive. And since this is really only possible (without patching the source) with this change, I find it has relevant to this PR. I don't care if it is another commit in this PR, but it is something we should have. Personally think we should have always have had and it could have use even without this PR, wouldn't be opposed to open a new PR for it, but don't think it is worth the trouble since I would consider this PR to be dependent on that. Do you have some objection to it? If not then would just making it a seperate commit suffice?

The mode `MODE_ALL = -1` is actually kinda important to the end user. Yes, I had made it -1 because -1 had the same effect as the enum before this change, but making it not 0 was intentional. If add add a new keybinding without specifying the mode, it'll default to 0. So whichever enum value as the value of 0 will be the default in that case. Not saying that `MODE_IMAGE` is completely the right answer, but I feel like you just accidentally changed the default. Moving MODE_ALL to the end of the enum would also work. So my question to you is what should the default mode be when not specified? If we want MODE_ALL to be the default, that that change is fine. To me MODE_IMAGE is a good default, but I don't have anything close to a objective argument to offer.


> Is there any benefit of guarding `command.h` with the `#ifdef` ?

Feel like it is a good practice to always guard header files. May not be much practical benefit in this small code base. It could have some if we do something fancier in commands.h like storing a mapping of func name to pointer for use of IPCs or such.

N-R-K commented:
> If not then would just making it a seperate commit suffice?

Yeah, I think I said the same thing [here](https://github.com/nsxiv/nsxiv/pull/76#discussion_r720199093)

>  I feel like you just accidentally changed the default.

At the time I wrote my msg, yes. But I realized that when reading your initial commit msg, didn't update/edit my comment since MODE_ALL default doesn't seemed like a better default.

TAAPArthur commented:
> Yeah, I think I said the same thing [here](https://github.com/nsxiv/nsxiv/pull/76#discussion_r720199093)

Missed your comment. My bad.

>At the time I wrote my msg, yes. But I realized that when reading your initial commit msg, didn't update/edit my comment since MODE_ALL default seem like a better default.

Sounds good. I can see it both ways and think anyone who cares should be be explicit. Also think it is better than using -1 even if it is just cosmetic.

Kept all the changes you suggested, squashed everything save for the imgcursor and rebased on master

N-R-K commented:
Added comment to imgcursor, made it `static const` and fixed the commit msg (changed default to MODE_ALL).

N-R-K commented:
@TAAPArthur was just waiting for a 2nd approval. I approved it long ago.
